In this project, we'll work with a dataset of used cars from German eBay.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

It looks like the columns are all following a camel case with the second/third words being capitalized. We'll need to correct column names. There also appear to be some missing values in 4 columns. Several columns also appear to be of a different dtype than expected (dateCrawled, price, yearOfRegistration, odometer, monthOfRegistration, lastSeen).

We'll first rename some of the columns and change them to snake case.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

We'll next do some basic data exploration to see whether more data cleaning is required. Specifically, we'll look for columns where all or almost all values are the same, as they can be dropped for not having useful information for analysis. We'll also look at numeric data stored in as text, which can be cleaned and converted.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

## Issues at first glance

Any columns that have mostly one value that are candidates to be dropped 
- seller, offer_type? abtest?

Any columns that need more investigation 

- price $0 appears 1421 times, registration_year max 9999, min 1000

Initial observations:
- There are a number of columns where all (or almost all) the values are the same
- seller, offer_type
- The num_photos column looks odd. We'll need to investigate further.

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the column num_photos has 0 as a value for all its rows. Thus, we'll drop it.

In [7]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

---

We'll start with columns 'price' and 'odometer' as two of the several columns with numeric data stored in as text.

In [8]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object

In [9]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos.rename(columns={'price' : 'price_usd'}, inplace=True)

In [10]:
autos['price_usd']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price_usd, Length: 50000, dtype: int64

We'll next clean up 'odometer' column. Let's see what it looks like.

In [11]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

In [13]:
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64

---

We'll now further explore 'price' and 'odometer_km' columns, as these will be the ones we'll be doing our analysis on.

In [14]:
autos['price_usd'].unique().shape

(2357,)

In [15]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

Max price seems to be too much (100000000)

In [16]:
autos['price_usd'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price_usd, Length: 2357, dtype: int64

In [17]:
autos['price_usd'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

It appears that there are several cars with an unrealistic price for a used model. We'll now remove outliers in the 'price_usd' column. Upon further inspection, it looks like the highest most probable price is 350000 for a 2016 Porsche 911, although there are two Ferraris one sold for 1.3m and the other for 3.9m. Anything with a price higher than 350000 is considered an outlier and will be removed.

There are also multiple cars with a price below 30 with almost 1500 priced at 0. But since eBay is an auction site, it makes sense that the starting bid for some cars begins at 0.

In [18]:
autos.loc[autos['price_usd'] > 350000].head()

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   

       price_usd   abtest vehicle_type  registration_year    gearbox  \
514       999999     test        kombi               2009    manuell   
2897    11111111     test    limousine               1973    manuell   
7814     1300000  control        coupe               1992        NaN   
11137   10000000  control        coupe               1960    manuell   
22947    1234566  control        kombi               1999  automatik   

       power_ps   model  odometer_km  registration_month fuel_type  \
514         101   focus       125000                   4    benzin   
2897         48  escort        50000                   3    benzin   
7814          0     NaN        50000                  12       NaN   
11137       368     NaN       100000                   1    benzin   
22947       190     NaN       150000                   2    diesel   

                brand unrepaired_damage           ad_created  postal_code  \
514              ford              nein  2016-03-17 00:00:00        12205   
2897             ford              nein  2016-03-12 00:00:00        94469   
7814   sonstige_autos              nein  2016-04-04 00:00:00        60598   
11137  sonstige_autos              nein  2016-03-29 00:00:00        73033   
22947             bmw               NaN  2016-03-22 00:00:00        17454   

                 last_seen  
514    2016-04-06 07:17:35  
2897   2016-03-12 22:45:27  
7814   2016-04-05 11:34:11  
11137  2016-04-06 21:18:11  
22947  2016-04-02 03:17:32

In [19]:
autos = autos[autos['price_usd'] < 350001]

In [20]:
autos.shape

(49986, 17)

Let's now take a closer look at the 'odometer_km' column.

In [21]:
autos['odometer_km'].unique().shape

(13,)

In [22]:
autos['odometer_km'].describe()

count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
autos['odometer_km'].value_counts()

150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
5000        966
40000       818
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

Odometer data seems to be in the increments of 10000 until 100000, and increments of 25000 after that with 150000 being the highest possible odometer reading, which might mean that the seller had to choose from a pre-set rounded options for this field.

---

Let's now move on to the date columns and see what date range is covered. At the moment, there are 5 columns that represent dates: date_crawled, last_seen, ad_created, registration_month, registration_year. Let's use the describe( ) method to see the distribution of the numeric columns first.

In [24]:
autos[['date_crawled', 'last_seen', 'ad_created', 'registration_month', 'registration_year']].describe()

registration_month  registration_year
count        49986.000000       49986.000000
mean             5.723723        2005.075721
std              3.711839         105.727161
min              0.000000        1000.000000
25%              3.000000        1999.000000
50%              6.000000        2003.000000
75%              9.000000        2008.000000
max             12.000000        9999.000000

Of particular interest here is the registration_year column, with 1000 being the minimum year and 9999 being the maximum year. We'll come back to this later in this project.

We'll also take a look at the other three columns to understand how they are formatted.

In [25]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [26]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [27]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64

In [28]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

date_crawled and last_seen columns seem to have the same date distribution, beginning on 03/05/2016 and ending on 04/07/2016, whereas ad_created column starts on 06/11/2015.

Let's now go back to registration_year column.

In [29]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

A car can't be first registered after the listing was seen, so any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult.

In [30]:
(~autos['registration_year'].between(1900, 2016)).sum()

1970

In [31]:
(~autos['registration_year'].between(1900, 2016)).sum() / autos.shape[0]

0.03941103508982515

Given that autos registered earlier than 1900 and after 2016 constitute only 4% of the data, we can remove those rows.

In [32]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
2006    0.056377
2001    0.056273
2002    0.052753
1998    0.051087
2007    0.047984
Name: registration_year, dtype: float64

It appears most of the cars were first registered in the last 20 or so years.

## Exploring price by brand

In [33]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
smart             0.013912
citroen           0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

This data being from a German eBay, German cars represent 4 of the top 5 brands, with about 50% of all listings. VW is the most popular brand with almost double the number of listing compared to the second most popular brand, BMW. 

We'll limit our analysis to brands with more than 5% of the total listings, as the rest is not very significant.

In [34]:
brand_counts = autos['brand'].value_counts(normalize=True)
popular_brands = brand_counts[brand_counts > .05].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [35]:
brand_mean_prices = {}

for brand in popular_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price_usd'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5231,
 'bmw': 8102,
 'opel': 2876,
 'mercedes_benz': 8485,
 'audi': 9093,
 'ford': 3652}

Of the top brands, Audi, Mercedes Benz and BMW are the most expensive. Ford and Open are the least expensive, while VW is in-between, which might explain its popularity.

## Exploring mileage

In [36]:
brand_mean_mileage = {}

for brand in popular_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

brand_mean_mileage

{'volkswagen': 128724,
 'bmw': 132431,
 'opel': 129223,
 'mercedes_benz': 130856,
 'audi': 129287,
 'ford': 124068}

In [37]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [38]:
brand_info = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132431
mercedes_benz        130856
audi                 129287
opel                 129223
volkswagen           128724
ford                 124068

In [39]:
brand_info['mean_prices'] = mean_prices
brand_info

mean_mileage  mean_prices
bmw                  132431         8102
mercedes_benz        130856         8485
audi                 129287         9093
opel                 129223         2876
volkswagen           128724         5231
ford                 124068         3652

There is not a very big variation in mileage, but it does seem like more expensive cars have higher mileage on average and less expensive car have lower mileage on average.

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?